File: OpenVehicleApp/src/com/openvehicles/OVMS/ui/settings/CarInfoFragment.java
package com.openvehicles.OVMS.ui.settings;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager.NameNotFoundException;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.actionbarsherlock.app.SherlockFragmentActivity;
import com.openvehicles.OVMS.R;
import com.openvehicles.OVMS.entities.CarData;
import com.openvehicles.OVMS.ui.BaseFragment;
import com.openvehicles.OVMS.ui.utils.Ui;
import com.openvehicles.OVMS.utils.CarsStorage;

public class CarInfoFragment extends BaseFragment {
	private CarData mCarData;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(R.layout.fragment_car_info, null);
	}
	
	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		SherlockFragmentActivity activity = getSherlockActivity(); 
		
		activity.getSupportActionBar().setIcon(R.drawable.ic_action_about);
		activity.setTitle(R.string.lb_vehicle_info);
		
		int editPosition = getArguments().getInt("position", -1);
		if (editPosition >= 0) {
			mCarData = CarsStorage.get().getStoredCars().get(editPosition);
			
		}
		
		if (mCarData != null) approveCarData();
	}
	
	private void approveCarData() {
		View rootView = getView();
		Context context = rootView.getContext();
		
		Ui.setValue(rootView, R.id.txt_vehicle_id, mCarData.sel_vehicleid);
		Ui.setValue(rootView, R.id.txt_win, mCarData.car_vin);
		Ui.setValue(rootView, R.id.txt_type, mCarData.car_type);
		Ui.setValue(rootView, R.id.txt_server, mCarData.server_firmware);
		Ui.setValue(rootView, R.id.txt_car, mCarData.car_firmware);
		Ui.setValue(rootView, R.id.txt_gsm, mCarData.car_gsm_signal);
		Ui.setValue(rootView, R.id.txt_cac, String.format("%.2f",mCarData.car_CAC));

		Ui.setValue(rootView, R.id.txt_12v_info, String.format("%.1fV (%s)",
			mCarData.car_12vline_voltage,
			mCarData.car_charging_12v
				? "charging"
				: (mCarData.car_12vline_ref <= 1.5)
					? String.format("calmdown, %d min left", 15 - (int)(mCarData.car_12vline_ref*10))
					: String.format("ref=%.1fV", mCarData.car_12vline_ref)));

		ImageView iv = (ImageView)rootView.findViewById(R.id.img_signal_rssi);
		iv.setImageResource(Ui.getDrawableIdentifier(context, "signal_strength_" + mCarData.car_gsm_bars));
		
		try {
			PackageInfo pi = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
			Ui.setValue(rootView, R.id.txt_app, String.format("%s (%s)", pi.versionName, pi.versionCode));
		} catch (NameNotFoundException e) { }
		
	}
	
	@Override
	public void update(CarData pCarData) {
		mCarData = pCarData;
		approveCarData();
	}
}

File: app/src/main/java/com/openvehicles/OVMS/ui/settings/CarInfoFragment.kt
package com.openvehicles.OVMS.ui.settings

import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.openvehicles.OVMS.R
import com.openvehicles.OVMS.entities.CarData
import com.openvehicles.OVMS.ui.BaseFragment
import com.openvehicles.OVMS.ui.utils.Ui.getDrawableIdentifier
import com.openvehicles.OVMS.ui.utils.Ui.setValue
import com.openvehicles.OVMS.utils.CarsStorage.getSelectedCarData
import com.openvehicles.OVMS.utils.CarsStorage.getStoredCars

class CarInfoFragment : BaseFragment() {

    private var carData: CarData? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_car_info, null)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        compatActivity?.setTitle(R.string.lb_vehicle_info)
        val editPosition = requireArguments().getInt("position", -1)
        carData = if (editPosition >= 0) {
            getStoredCars()[editPosition]
        } else {
            getSelectedCarData()
        }
        if (carData != null) {
            approveCarData()
        }
    }

    private fun approveCarData() {
        val rootView = view
        val context = rootView!!.context
        setValue(rootView, R.id.txt_vehicle_id, carData!!.sel_vehicleid)
        setValue(rootView, R.id.txt_win, carData!!.car_vin)
        setValue(rootView, R.id.txt_type, carData!!.car_type)
        setValue(rootView, R.id.txt_server, carData!!.server_firmware)
        setValue(rootView, R.id.txt_car, carData!!.car_firmware)
        setValue(rootView, R.id.txt_hardware, carData!!.car_hardware)
        setValue(rootView,
            R.id.txt_gsm,
            String.format(
                "%s  %s  %s",carData!!.car_gsm_signal, carData!!.car_gsmlock, carData!!.car_mdm_mode)
        )
        setValue(
            rootView,
            R.id.txt_cac,
            if (carData!!.car_CAC_percent > 0) String.format(
                "%.1f Ah = %.1f%%",
                carData!!.car_CAC,
                carData!!.car_CAC_percent
            ) else if (carData!!.car_type == "SQ") String.format(
                "%.1f Ah   SoC: %.0f%%  %.1fkWh",
                carData!!.car_CAC,
                carData!!.car_soc_raw,
                carData!!.car_battery_capacity
            ) else String.format("%.1f Ah", carData!!.car_CAC)
        )
        setValue(rootView, R.id.txt_soh, String.format("%.1f%%", carData!!.car_soh))
        setValue(
            rootView, R.id.txt_12v_info, String.format(
                "%.1fV (%s) %.1fA",
                carData!!.car_12vline_voltage,
                if (carData!!.car_charging_12v) "charging" else if (carData!!.car_12vline_ref <= 1.5) String.format(
                    "calmdown, %d min left",
                    15 - (carData!!.car_12vline_ref * 10).toInt()
                ) else String.format("ref=%.1fV", carData!!.car_12vline_ref),
                carData!!.car_12v_current
            )
        )
        setValue(
            rootView,
            R.id.txt_charge_info,
            String.format("%.1f kWh", carData!!.car_charge_kwhconsumed)
        )

        // Show known car service interval info:
        var serviceInfo = ""
        if (carData!!.car_servicerange >= 0) {
            serviceInfo += String.format("%d km", carData!!.car_servicerange)
        }
        if (carData!!.car_servicetime >= 0) {
            if (serviceInfo != "") {
                serviceInfo += " / "
            }
            val now = System.currentTimeMillis() / 1000
            val serviceDays = (carData!!.car_servicetime - now) / 86400
            serviceInfo += String.format("%d days", serviceDays)
        }
        val serviceTextView = rootView.findViewById<View>(R.id.txt_service_info) as TextView
        val serviceView = rootView.findViewById<View>(R.id.service_info) as TextView
        if (serviceInfo == "") {
            serviceView.visibility = View.GONE
            serviceTextView.visibility = View.GONE
        } else {
            serviceView.visibility = View.VISIBLE
            serviceTextView.visibility = View.VISIBLE
            serviceTextView.text = serviceInfo
        }
        val iv = rootView.findViewById<View>(R.id.img_signal_rssi) as ImageView
        iv.setImageResource(
            getDrawableIdentifier(
                context,
                "signal_strength_" + carData!!.car_gsm_bars
            )
        )

        try {
            val pi = context.packageManager.getPackageInfo(context.packageName, 0)
            setValue(
                rootView,
                R.id.txt_app,
                String.format("%s (%s)", pi.versionName, pi.versionCode)
            )
        } catch (e: PackageManager.NameNotFoundException) {
        }
    }

    override fun update(carData: CarData?) {
        this.carData = carData
        approveCarData()
    }
}
